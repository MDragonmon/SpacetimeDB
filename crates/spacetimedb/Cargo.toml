[package]
name = "spacetimedb"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[[bin]]
name = "spacetimedb"   # The name of the target.
path = "src/main.rs"   # The source file of the target.
test = true            # Is tested by default.
bench = true           # Is benchmarked by default.
doc = false            # Is documented by default.
plugin = false         # Used as a compiler plugin (deprecated).
proc-macro = false     # Set to `true` for a proc-macro library.
harness = true         # Use libtest harness.
required-features = [] # Features required to build this target (N/A for lib)

[[bench]]
name = "transactional_db"
harness = false

[[bench]]
name = "object_db"
harness = false

[dependencies] 
tokio = { version = "1.2", features = ["full"] }
wasmer = "2.1.1"
wasm-bindgen = "0.2.79"
wasmer-middlewares = "2.1.1"
sha3 = "0.10.0"
hex-literal = "0.3.4"
hex = "0.4.3"
protobuf = { version = "2" }
prost = "0.6"
lazy_static = "1.4.0"
# wasmer-compiler-llvm = "2.2.1"
spacetimedb-bindings = { path = "../spacetimedb-bindings" }
log = "0.4.16"
gotham = "0.7.1"
gotham_derive = "0.7"
hyper = "0.14.18"
log4rs = { version = "1.0.0", features = ["background_rotation", "rolling_file_appender", "compound_policy", "size_trigger", "fixed_window_roller"] }
serde = "1.0.136"
serde_json = "1.0"
anyhow = { version = "1.0.57", features = ["backtrace"] }
jsonwebtoken = { version = "8.1.0" }
sha1 = "0.10.1"
base64 = "0.13.0"
tokio-tungstenite = "*"
regex = "1"
futures = "0.3"
prometheus = "0.13.0"
hostname = "^0.3"
clap = { version = "3.2.8", features = ["derive"] }
sled = "0.34.7"
email_address = "0.2.3"
int-enum = "0.4.0"
pyo3 = { version = "0.17.0"}
sqlparser = "0.21.0"
backtrace = "0.3.66"
pprof = { version = "0.10", features = ["flamegraph"] }
urlencoding = "2.1.2"
perf_monitor = "0.2"

# gluesql = "0.10"
bytes = "1.2.0" # TODO(rdaum): there's also a copy/fork of this inside prost. hmph.

# Rocksdb ostorage backend, linked only if "rocksdb" feature enabled.
[dependencies.rocksdb]
version = "0.18.0"
default-features = false
optional = true
features = ["lz4"] # if we link bzip, we get multiple defs

[features]
# Optional storage engines.
odb_rocksdb = ["dep:rocksdb"]
odb_sled = []

[dev-dependencies]
criterion = { version = "0.3", features = ["async", "async_tokio", "html_reports"] }
prost-build = { version = "0.6.1" }
tempdir = "0.3.7"
rand = "0.8.5"

[build-dependencies]
glob = "0.3.0"
prost-build = { version = "0.6.1" }
