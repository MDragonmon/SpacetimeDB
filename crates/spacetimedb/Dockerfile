# TODO maybe a multistage build eventually?
FROM rust:1.57.0
WORKDIR /usr/src/app

RUN cargo install cargo-watch

RUN apt update
RUN apt install -y lsb-release wget software-properties-common
RUN wget https://apt.llvm.org/llvm.sh
RUN chmod +x llvm.sh
RUN ./llvm.sh 12

# RUN apt update
# RUN apt install -y software-properties-common
# RUN wget --no-check-certificate -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
# RUN apt-add-repository "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-12 main"
# COPY ./crates/spacetimedb/libffi6_3.2.1-9_arm64.deb ./
# RUN apt-get install -y ./libffi6_3.2.1-9_arm64.deb
# RUN apt-get update && apt-get install -y libllvm12 llvm-12 llvm-12-dev llvm-12-doc llvm-12-examples llvm-12-runtime
RUN apt-get update && apt-get install -y \
  cmake \
  less 

COPY ./crates/spacetimedb-bindings ../spacetimedb-bindings
COPY ./crates/spacetimedb/src ./src
COPY ./crates/spacetimedb/protobuf ./protobuf

COPY ./Cargo.lock ./
COPY ./crates/spacetimedb/Cargo.toml ./

RUN echo "fn main() {}" > dummy.rs
RUN sed -i 's#src/main.rs#dummy.rs#' Cargo.toml
RUN cargo install --locked --path .
RUN sed -i 's#dummy.rs#src/main.rs#' Cargo.toml

COPY ./crates/spacetimedb ./

RUN cargo install --locked --path .

EXPOSE 3000

WORKDIR /usr/src/app
ENV RUST_BACKTRACE=1
CMD ["../target/release/spacetimedb"]
