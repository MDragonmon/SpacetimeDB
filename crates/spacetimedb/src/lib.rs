pub mod api;
pub mod db;
pub mod hash;
mod messages;
pub mod routes;
pub mod wasm_host;
pub mod postgres;
