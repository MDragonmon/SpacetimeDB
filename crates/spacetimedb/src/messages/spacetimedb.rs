#[derive(Clone, PartialEq, ::prost::Message)]
pub struct TableWrite {
    #[prost(uint32, tag="1")]
    pub table_id: u32,
    #[prost(message, repeated, tag="2")]
    pub writes: ::std::vec::Vec<Write>,
}
#[derive(Clone, PartialEq, ::prost::Message)]
pub struct Insert {
    #[prost(oneof="insert::Insert", tags="1, 2")]
    pub insert: ::std::option::Option<insert::Insert>,
}
pub mod insert {
    #[derive(Clone, PartialEq, ::prost::Oneof)]
    pub enum Insert {
        #[prost(bytes, tag="1")]
        Hash(std::vec::Vec<u8>),
        #[prost(bytes, tag="2")]
        Raw(std::vec::Vec<u8>),
    }
}
#[derive(Clone, PartialEq, ::prost::Message)]
pub struct Write {
    #[prost(oneof="write::Operation", tags="1, 2")]
    pub operation: ::std::option::Option<write::Operation>,
}
pub mod write {
    #[derive(Clone, PartialEq, ::prost::Oneof)]
    pub enum Operation {
        #[prost(message, tag="1")]
        Insert(super::Insert),
        #[prost(bytes, tag="2")]
        Delete(std::vec::Vec<u8>),
    }
}
#[derive(Clone, PartialEq, ::prost::Message)]
pub struct Commit {
    #[prost(bytes, tag="1")]
    pub parent_commit_hash: std::vec::Vec<u8>,
    #[prost(message, repeated, tag="2")]
    pub table_writes: ::std::vec::Vec<TableWrite>,
}
