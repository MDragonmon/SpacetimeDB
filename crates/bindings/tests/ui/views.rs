use spacetimedb::{reducer, table, view, AnonymousViewContext, Identity, ReducerContext, ViewContext};

#[table(name = test)]
struct Test {
    #[unique]
    id: u32,
    #[index(btree)]
    x: u32,
}

#[reducer]
fn view_handle_no_iter(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `iter()`
    for _ in read_only.db.test().iter() {}
}

#[reducer]
fn view_handle_no_count(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `count()`
    let _ = read_only.db.test().count();
}

#[reducer]
fn view_handle_no_insert(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `insert()`
    read_only.db.test().insert(Test { id: 0, x: 0 });
}

#[reducer]
fn view_handle_no_try_insert(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `try_insert()`
    read_only.db.test().try_insert(Test { id: 0, x: 0 });
}

#[reducer]
fn view_handle_no_delete(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: ViewHandle does not expose `delete()`
    read_only.db.test().delete(Test { id: 0, x: 0 });
}

#[reducer]
fn read_only_unqiue_index_no_delete(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: unique read-only index does not expose `delete()`
    read_only.db.test().id().delete(&0);
}

#[reducer]
fn read_only_unqiue_index_no_update(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: unique read-only index does not expose `update()`
    read_only.db.test().id().update(Test { id: 0, x: 0 });
}

#[reducer]
fn read_only_btree_index_no_delete(ctx: &ReducerContext) {
    let read_only = ctx.as_read_only();
    // Should not compile: read-only btree index does not expose `delete()`
    read_only.db.test().x().delete(0u32..);
}

#[table(name = player)]
struct Player {
    #[unique]
    identity: Identity,
}

/// Private views not allowed; must be `#[view(public)]`
#[view]
fn view_def_no_public(_: &ViewContext) -> Vec<Player> {
    vec![]
}

/// A `ViewContext` is required
#[view(public)]
fn view_def_no_context() -> Vec<Player> {
    vec![]
}

/// A `ViewContext` is required
#[view(public)]
fn view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
    vec![]
}

/// A `ViewContext` is required
#[view(public)]
fn view_def_wrong_context_2(_: &AnonymousViewContext) -> Vec<Player> {
    vec![]
}

/// An `AnonymousViewContext` is required
#[view(public, anonymous)]
fn anonymous_view_def_no_context() -> Vec<Player> {
    vec![]
}

/// An `AnonymousViewContext` is required
#[view(public, anonymous)]
fn anonymous_view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
    vec![]
}

/// An `AnonymousViewContext` is required
#[view(public, anonymous)]
fn anonymous_view_def_wrong_context_2(_: &ViewContext) -> Vec<Player> {
    vec![]
}

/// Must return `Vec<T>` where `T` is a SpacetimeType
#[view(public)]
fn view_def_no_return(_: &ViewContext) {}

/// Must return `Vec<T>` where `T` is a SpacetimeType
#[view(public)]
fn view_def_wrong_return(_: &ViewContext) -> Option<Player> {
    None
}

fn main() {}
