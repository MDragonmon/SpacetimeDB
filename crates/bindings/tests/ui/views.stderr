error: views must be declared as #[view(public)]; public is required
  --> tests/ui/views.rs:74:1
   |
74 | #[view]
   | ^^^^^^^
   |
   = note: this error originates in the attribute macro `view` (in Nightly builds, run with -Z macro-backtrace for more info)

error: A view must have `&ViewContext` as its first argument
  --> tests/ui/views.rs:81:1
   |
81 | fn view_def_no_context() -> Vec<Player> {
   | ^^

error: An anonymous view must have `&AnonymousViewContext` as its first argument
  --> tests/ui/views.rs:99:1
   |
99 | fn anonymous_view_def_no_context() -> Vec<Player> {
   | ^^

error: views must return `Vec<T>` where `T` is a `SpacetimeType`
   --> tests/ui/views.rs:117:1
    |
117 | fn view_def_no_return(_: &ViewContext) {}
    | ^^

error[E0599]: no method named `iter` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:15:34
   |
15 |     for _ in read_only.db.test().iter() {}
   |                                  ^^^^ method not found in `&test__ViewHandle`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `iter`, perhaps you need to implement one of them:
           candidate #1: `Table`
           candidate #2: `bitflags::traits::Flags`

error[E0599]: `&test__ViewHandle` is not an iterator
  --> tests/ui/views.rs:22:33
   |
 3 | #[table(name = test)]
   | ------------------- doesn't satisfy `test__ViewHandle: Iterator`
...
22 |     let _ = read_only.db.test().count();
   |                                 ^^^^^ `&test__ViewHandle` is not an iterator
   |
   = note: the following trait bounds were not satisfied:
           `&test__ViewHandle: Iterator`
           which is required by `&mut &test__ViewHandle: Iterator`
           `test__ViewHandle: Iterator`
           which is required by `&mut test__ViewHandle: Iterator`
note: the trait `Iterator` must be implemented
  --> $RUST/core/src/iter/traits/iterator.rs
   |
   | pub trait Iterator {
   | ^^^^^^^^^^^^^^^^^^
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `count`, perhaps you need to implement one of them:
           candidate #1: `Table`
   = note: the trait `Iterator` defines an item `count`, but is explicitly unimplemented

error[E0599]: no method named `insert` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:29:25
   |
29 |     read_only.db.test().insert(Test { id: 0, x: 0 });
   |                         ^^^^^^ method not found in `&test__ViewHandle`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following traits define an item `insert`, perhaps you need to implement one of them:
           candidate #1: `Table`
           candidate #2: `bitflags::traits::Flags`
           candidate #3: `ppv_lite86::types::Vec2`
           candidate #4: `ppv_lite86::types::Vec4`

error[E0599]: no method named `try_insert` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:36:25
   |
36 |     read_only.db.test().try_insert(Test { id: 0, x: 0 });
   |                         ^^^^^^^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `try_insert`, perhaps you need to implement it:
           candidate #1: `Table`
help: there is a method `try_into` with a similar name, but with different arguments
  --> $RUST/core/src/convert/mod.rs
   |
   |     fn try_into(self) -> Result<T, Self::Error>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `delete` found for reference `&test__ViewHandle` in the current scope
  --> tests/ui/views.rs:43:25
   |
43 |     read_only.db.test().delete(Test { id: 0, x: 0 });
   |                         ^^^^^^ method not found in `&test__ViewHandle`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `delete`, perhaps you need to implement it:
           candidate #1: `Table`

error[E0599]: no method named `delete` found for struct `UniqueColumnReadOnly` in the current scope
  --> tests/ui/views.rs:50:30
   |
50 |     read_only.db.test().id().delete(&0);
   |                              ^^^^^^ method not found in `UniqueColumnReadOnly<test__TableHandle, u32, id>`

error[E0599]: no method named `update` found for struct `UniqueColumnReadOnly` in the current scope
  --> tests/ui/views.rs:57:30
   |
57 |     read_only.db.test().id().update(Test { id: 0, x: 0 });
   |                              ^^^^^^ method not found in `UniqueColumnReadOnly<test__TableHandle, u32, id>`

error[E0599]: no method named `delete` found for struct `RangedIndexReadOnly` in the current scope
  --> tests/ui/views.rs:64:29
   |
64 |     read_only.db.test().x().delete(0u32..);
   |                             ^^^^^^ method not found in `RangedIndexReadOnly<test__TableHandle, (u32,), x>`

error[E0277]: invalid view signature
  --> tests/ui/views.rs:87:4
   |
86 | #[view(public)]
   | --------------- required by a bound introduced by this call
87 | fn view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
   |    ^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
   |
   = help: the trait `spacetimedb::rt::View<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext) -> Vec<Player> {view_def_wrong_context_1}`
   = note:
   = note: view signatures must match:
   = note:     `Fn(&ViewContext, [T1, ...]) -> Vec<Tn>`
   = note: where each `Ti` implements `SpacetimeType`.
   = note:
note: required by a bound in `register_view`
  --> src/rt.rs
   |
   | pub fn register_view<'a, A, I, T>(_: impl View<'a, A, T>)
   |                                           ^^^^^^^^^^^^^^ required by this bound in `register_view`

error[E0277]: A view must have `&ViewContext` as its first argument
  --> tests/ui/views.rs:87:32
   |
87 | fn view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
   |                                ^^^^^^^^^^^^^^^ the trait `ViewContextArg` is not implemented for `&ReducerContext`
   |
   = help: the trait `ViewContextArg` is implemented for `&ViewContext`

error[E0277]: invalid view signature
  --> tests/ui/views.rs:87:4
   |
86 | #[view(public)]
   | --------------- required by a bound introduced by this call
87 | fn view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
   |    ^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
   |
   = help: the trait `spacetimedb::rt::View<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext) -> Vec<Player> {view_def_wrong_context_1}`
   = note:
   = note: view signatures must match:
   = note:     `Fn(&ViewContext, [T1, ...]) -> Vec<Tn>`
   = note: where each `Ti` implements `SpacetimeType`.
   = note:
note: required by a bound in `invoke_view`
  --> src/rt.rs
   |
   | pub fn invoke_view<'a, A: Args<'a>, T: SpacetimeType + Serialize>(
   |        ----------- required by a bound in this function
   |     view: impl View<'a, A, T>,
   |                ^^^^^^^^^^^^^^ required by this bound in `invoke_view`

error[E0277]: invalid view signature
  --> tests/ui/views.rs:93:4
   |
92 | #[view(public)]
   | --------------- required by a bound introduced by this call
93 | fn view_def_wrong_context_2(_: &AnonymousViewContext) -> Vec<Player> {
   |    ^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
   |
   = help: the trait `spacetimedb::rt::View<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a AnonymousViewContext) -> Vec<Player> {view_def_wrong_context_2}`
   = note:
   = note: view signatures must match:
   = note:     `Fn(&ViewContext, [T1, ...]) -> Vec<Tn>`
   = note: where each `Ti` implements `SpacetimeType`.
   = note:
note: required by a bound in `register_view`
  --> src/rt.rs
   |
   | pub fn register_view<'a, A, I, T>(_: impl View<'a, A, T>)
   |                                           ^^^^^^^^^^^^^^ required by this bound in `register_view`

error[E0277]: A view must have `&ViewContext` as its first argument
  --> tests/ui/views.rs:93:32
   |
93 | fn view_def_wrong_context_2(_: &AnonymousViewContext) -> Vec<Player> {
   |                                ^^^^^^^^^^^^^^^^^^^^^ the trait `ViewContextArg` is not implemented for `&AnonymousViewContext`
   |
   = help: the trait `ViewContextArg` is implemented for `&ViewContext`

error[E0277]: invalid view signature
  --> tests/ui/views.rs:93:4
   |
92 | #[view(public)]
   | --------------- required by a bound introduced by this call
93 | fn view_def_wrong_context_2(_: &AnonymousViewContext) -> Vec<Player> {
   |    ^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
   |
   = help: the trait `spacetimedb::rt::View<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a AnonymousViewContext) -> Vec<Player> {view_def_wrong_context_2}`
   = note:
   = note: view signatures must match:
   = note:     `Fn(&ViewContext, [T1, ...]) -> Vec<Tn>`
   = note: where each `Ti` implements `SpacetimeType`.
   = note:
note: required by a bound in `invoke_view`
  --> src/rt.rs
   |
   | pub fn invoke_view<'a, A: Args<'a>, T: SpacetimeType + Serialize>(
   |        ----------- required by a bound in this function
   |     view: impl View<'a, A, T>,
   |                ^^^^^^^^^^^^^^ required by this bound in `invoke_view`

error[E0277]: invalid view signature
   --> tests/ui/views.rs:105:4
    |
104 | #[view(public, anonymous)]
    | -------------------------- required by a bound introduced by this call
105 | fn anonymous_view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
    |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
    |
    = help: the trait `AnonymousView<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext) -> Vec<Player> {anonymous_view_def_wrong_context_1}`
    = note:
    = note: view signatures must match:
    = note:     `Fn(&AnonymousViewContext, [T1, ...]) -> Vec<Tn>`
    = note: where each `Ti` implements `SpacetimeType`.
    = note:
note: required by a bound in `register_anonymous_view`
   --> src/rt.rs
    |
    | pub fn register_anonymous_view<'a, A, I, T>(_: impl AnonymousView<'a, A, T>)
    |                                                     ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `register_anonymous_view`

error[E0277]: An anonymous view must have `&AnonymousViewContext` as its first argument
   --> tests/ui/views.rs:105:42
    |
105 | fn anonymous_view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
    |                                          ^^^^^^^^^^^^^^^ the trait `AnonymousViewContextArg` is not implemented for `&ReducerContext`
    |
    = help: the trait `AnonymousViewContextArg` is implemented for `&AnonymousViewContext`

error[E0277]: invalid view signature
   --> tests/ui/views.rs:105:4
    |
104 | #[view(public, anonymous)]
    | -------------------------- required by a bound introduced by this call
105 | fn anonymous_view_def_wrong_context_1(_: &ReducerContext) -> Vec<Player> {
    |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
    |
    = help: the trait `AnonymousView<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ReducerContext) -> Vec<Player> {anonymous_view_def_wrong_context_1}`
    = note:
    = note: view signatures must match:
    = note:     `Fn(&AnonymousViewContext, [T1, ...]) -> Vec<Tn>`
    = note: where each `Ti` implements `SpacetimeType`.
    = note:
note: required by a bound in `invoke_anonymous_view`
   --> src/rt.rs
    |
    | pub fn invoke_anonymous_view<'a, A: Args<'a>, T: SpacetimeType + Serialize>(
    |        --------------------- required by a bound in this function
    |     view: impl AnonymousView<'a, A, T>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `invoke_anonymous_view`

error[E0277]: invalid view signature
   --> tests/ui/views.rs:111:4
    |
110 | #[view(public, anonymous)]
    | -------------------------- required by a bound introduced by this call
111 | fn anonymous_view_def_wrong_context_2(_: &ViewContext) -> Vec<Player> {
    |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
    |
    = help: the trait `AnonymousView<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ViewContext) -> Vec<Player> {anonymous_view_def_wrong_context_2}`
    = note:
    = note: view signatures must match:
    = note:     `Fn(&AnonymousViewContext, [T1, ...]) -> Vec<Tn>`
    = note: where each `Ti` implements `SpacetimeType`.
    = note:
note: required by a bound in `register_anonymous_view`
   --> src/rt.rs
    |
    | pub fn register_anonymous_view<'a, A, I, T>(_: impl AnonymousView<'a, A, T>)
    |                                                     ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `register_anonymous_view`

error[E0277]: An anonymous view must have `&AnonymousViewContext` as its first argument
   --> tests/ui/views.rs:111:42
    |
111 | fn anonymous_view_def_wrong_context_2(_: &ViewContext) -> Vec<Player> {
    |                                          ^^^^^^^^^^^^ the trait `AnonymousViewContextArg` is not implemented for `&ViewContext`
    |
    = help: the trait `AnonymousViewContextArg` is implemented for `&AnonymousViewContext`

error[E0277]: invalid view signature
   --> tests/ui/views.rs:111:4
    |
110 | #[view(public, anonymous)]
    | -------------------------- required by a bound introduced by this call
111 | fn anonymous_view_def_wrong_context_2(_: &ViewContext) -> Vec<Player> {
    |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
    |
    = help: the trait `AnonymousView<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ViewContext) -> Vec<Player> {anonymous_view_def_wrong_context_2}`
    = note:
    = note: view signatures must match:
    = note:     `Fn(&AnonymousViewContext, [T1, ...]) -> Vec<Tn>`
    = note: where each `Ti` implements `SpacetimeType`.
    = note:
note: required by a bound in `invoke_anonymous_view`
   --> src/rt.rs
    |
    | pub fn invoke_anonymous_view<'a, A: Args<'a>, T: SpacetimeType + Serialize>(
    |        --------------------- required by a bound in this function
    |     view: impl AnonymousView<'a, A, T>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `invoke_anonymous_view`

error[E0277]: invalid view signature
   --> tests/ui/views.rs:121:4
    |
120 | #[view(public)]
    | --------------- required by a bound introduced by this call
121 | fn view_def_wrong_return(_: &ViewContext) -> Option<Player> {
    |    ^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
    |
    = help: the trait `spacetimedb::rt::View<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ViewContext) -> Option<Player> {view_def_wrong_return}`
    = note:
    = note: view signatures must match:
    = note:     `Fn(&ViewContext, [T1, ...]) -> Vec<Tn>`
    = note: where each `Ti` implements `SpacetimeType`.
    = note:
note: required by a bound in `register_view`
   --> src/rt.rs
    |
    | pub fn register_view<'a, A, I, T>(_: impl View<'a, A, T>)
    |                                           ^^^^^^^^^^^^^^ required by this bound in `register_view`

error[E0277]: Views must return `Vec<T>` where `T` is a `SpacetimeType`
   --> tests/ui/views.rs:121:46
    |
121 | fn view_def_wrong_return(_: &ViewContext) -> Option<Player> {
    |                                              ^^^^^^^^^^^^^^ the trait `ViewReturn` is not implemented for `Option<Player>`
    |
    = help: the trait `ViewReturn` is implemented for `Vec<T>`

error[E0277]: invalid view signature
   --> tests/ui/views.rs:121:4
    |
120 | #[view(public)]
    | --------------- required by a bound introduced by this call
121 | fn view_def_wrong_return(_: &ViewContext) -> Option<Player> {
    |    ^^^^^^^^^^^^^^^^^^^^^ this view signature is not valid
    |
    = help: the trait `spacetimedb::rt::View<'_, _, _>` is not implemented for fn item `for<'a> fn(&'a ViewContext) -> Option<Player> {view_def_wrong_return}`
    = note:
    = note: view signatures must match:
    = note:     `Fn(&ViewContext, [T1, ...]) -> Vec<Tn>`
    = note: where each `Ti` implements `SpacetimeType`.
    = note:
note: required by a bound in `invoke_view`
   --> src/rt.rs
    |
    | pub fn invoke_view<'a, A: Args<'a>, T: SpacetimeType + Serialize>(
    |        ----------- required by a bound in this function
    |     view: impl View<'a, A, T>,
    |                ^^^^^^^^^^^^^^ required by this bound in `invoke_view`
